common/meta/api/src/schema_api_impl.rs

```diff
@@ -192,6 +194,80 @@ impl<KV: KVApi> SchemaApi for KV {
         }
     }
 
+    async fn rename_database(
+        &self,
+        req: RenameDatabaseReq,
+    ) -> Result<RenameDatabaseReply, MetaError> {
+        let tenant_dbname = &req.name_ident;
+        let tenant_newdbname = DatabaseNameIdent {
+            tenant: tenant_dbname.tenant.clone(),
+            db_name: req.new_db_name.clone(),
+        };
+
+        loop {
+            // get old db, not exists return err
+            let res = get_db_or_err(
+                self,
+                tenant_dbname,
+                format!("rename_database: {}", &tenant_dbname),
+            )
+            .await;
+
+            let (_, old_db_id, old_db_meta_seq, _) = match res {
+                Ok(x) => x,
+                Err(e) => {
+                    if let MetaError::AppError(AppError::UnknownDatabase(_)) = e {
+                        if req.if_exists {
+                            return Ok(RenameDatabaseReply {});
+                        }
+                    }
+                    return Err(e);
+                }
+            };
+
+            tracing::debug!(
+                old_db_id,
+                tenant_dbname = debug(&tenant_dbname),
+                "rename_database"
+            );
+            let old_id_key = DatabaseId { db_id: old_db_id };
+
+            // get new db, exists return err
+            let (db_id_seq, _db_id) = get_id_value(self, &tenant_newdbname).await?;
+            db_has_to_not_exist(db_id_seq, &tenant_newdbname, "rename_database")?;
+
+            // rename database
+            {
+                let txn_req = TxnRequest {
+                    condition: vec![
+                        // db has not to change. Renaming db is OK and does not affect the seq of db_meta.
+                        txn_cond_seq(&old_id_key, Eq, old_db_meta_seq)?,
```

Comment:
I think here it should assert `tenant_dbname, Eq, old_db_id_seq`, to prevent renaming or deleting in other threads.